### PR TITLE
Code appendix

### PR DESCRIPTION
**Pull request by kbjarkefur**
Divide up into a first section that is now a language agnostic part of what is good code

Then the follwoing two sections ( how to use our code examples & style guide) is all Stata

### CODE REVIEW

File: auxiliary/stata-guide.tex [hunk 1]
```diff
@@ -21,61 +21,27 @@
 people working in quantitative development research must be proficient collaborative programmers,
 and that includes more than being able to compute the correct numbers.
 
-This appendix begins with a short section containing instructions
+This appendix begins with a section on some general and language agnostic
+principles on how to write ``good'' code. 
+Good code is code that both generates the correct result \textit{and} 
+is easily read by other professionals in our field.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:27]:
```suggestion
is easily read by other professionals.
```

File: auxiliary/stata-guide.tex [hunk 2]
```diff
@@ -21,61 +21,27 @@
...
-\url{https://raw.githubusercontent.com/worldbank/d4di/master/code/code.do}.
-There, you can copy the code from your browser window to your do-file editor with the formatting intact.
-This method is only practical for a single file at the time.
-If you want to download all code used in this book, you can do that at:
-\url{https://github.com/worldbank/d4di/archive/master.zip}. That link offers a \texttt{.zip} file download
-with all the content used in writing this book, including the \LaTeX{} code used for the book itself. After
-extracting the .zip-file you will find all the code in a folder called \texttt{/code/}.
-
-\subsection{Writing and Understanding Stata code}
+\section{Writing code that is both correct and good}
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:44]:
```suggestion
\section{Writing good code}
```

It's odd to say both here in the title and then the first line says "good code is correct" I think

File: auxiliary/stata-guide.tex [hunk 3]
```diff
@@ -90,7 +56,7 @@ \subsection{Writing and Understanding Stata code}
 You should think of code in terms of three major elements:
 \textbf{structure}, \textbf{syntax}, and \textbf{style}.
 We always tell people to ``code as if a stranger would read it''
-(from tomorrow, that stranger could be you!).
+(from tomorrow, that stranger is most likely the future you!).
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:59]:
```suggestion
(and, tomorrow, we are strangers even to our past selves).
```

File: auxiliary/stata-guide.tex [hunk 4]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
 (or much less) accessible to someone who is reading it for the first time
 and needs to understand it quickly and correctly.
 
-Whether you are new to Stata or have used it for decades,
+Whether you are new to the programming language you are suing,
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:75]:
```suggestion
Whether you are new to the programming language you are using,
```

File: auxiliary/stata-guide.tex [hunk 5]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+(Stata, R, Python etc.)
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:76]:
```suggestion
(Stata, R, Python, or one of the many others)
```

File: auxiliary/stata-guide.tex [hunk 6]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+or have used it for decades,
 you will always run into commands that
 you have not seen before or whose function you do not remember.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:79]:
```suggestion
you have not seen before or whose functionality you do not remember.
```

File: auxiliary/stata-guide.tex [hunk 7]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
-(Whether you are new or not, you should frequently revisit the most common commands --
-often you will learn they can do something you never realized.\sidenote{
-  \url{https://www.stata.com/manuals13/u27.pdf}})
 Every time that happens,
 you should always look up the help file for that command.
 We often encounter the conception that help files are only for beginners.
 We could not disagree with that conception more,
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:83]:
```suggestion
We could not disagree with that conception more.
```

File: auxiliary/stata-guide.tex [hunk 8]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
-as the only way to get better at Stata is to constantly read help files.
-So if there is a command that you do not understand in any of our code examples,
-for example \texttt{isid}, then write \texttt{help isid},
-and the help file for the command \texttt{isid} will open.
+as the only way to get better at the programming language you use
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:84]:
```suggestion
The only way to get better at the programming language you use
```

File: auxiliary/stata-guide.tex [hunk 9]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+is to constantly read help files.
+(Whether you are new or not, you should frequently revisit the most common commands --
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:86]:
```suggestion
Whether you are new or not, you should frequently revisit the most common commands --
```

File: auxiliary/stata-guide.tex [hunk 10]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+often you will learn they can do something you never realized.)
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:87]:
```suggestion
often you will learn they can do something you never realized.
```

File: auxiliary/stata-guide.tex [hunk 11]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
 We cannot emphasize enough how important it is
 that you get into the habit of reading help files.
 Most of us have a help file window open at all times.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:91]: This section ends a bit abruptly, needs some sort of conclusion. I will take a stab at it.

File: auxiliary/stata-guide.tex [hunk 12]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
 
-Sometimes, you will encounter code that employs user-written commands,
-and you will not be able to read their help files until you have installed the commands.
-Two examples of these in our code are \texttt{randtreat} or \texttt{ieboilstart}.
+
+\section{Using the code examples in this book}
+
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:94]: A little intro needed here as well.

File: auxiliary/stata-guide.tex [hunk 13]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+We have ensured that each code block runs independently,
+is well-formatted, 
+and uses built-in commands as much as possible.
+We hope that these snippets will provide a foundation for your code style.
+Providing some standardization to Stata code style is also a goal of our team.
+Our code examples are commented generously (as you should),
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:100]:
```suggestion
We try to comment code examples generously (as you should),
```

File: auxiliary/stata-guide.tex [hunk 14]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+but you should reference Stata help-files by writing \texttt{help [command]}
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:101]:
```suggestion
and you should reference Stata help files by writing \texttt{help [command]}
```

"Help files" is Stata style, never "helpfiles" or "help-files"

File: auxiliary/stata-guide.tex [hunk 15]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+whenever you do not understand the command that is being used.
+For example, if you are not familiar with \texttt{isid},
+ then write \texttt{help isid},
+and the help file for the command \texttt{isid} will open.
+You should do this even if you know \texttt{isid} 
+but has not read the help file for that command in a while,
+as there is always something new to learn.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:108]: This section is a bit redundant with the previous one.

File: auxiliary/stata-guide.tex [hunk 16]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+
+In the book, code examples are presented like the following:
+
+\codeexample{code.do}{./code/code.do}
+
+You can access the raw code used in examples in this book in several ways.
+We use GitHub to version control all content of this book,
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:115]:
```suggestion
We use GitHub to version control all the content of this book,
```

File: auxiliary/stata-guide.tex [hunk 17]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+including the code examples.
+To see the code examples on GitHub, go to: \url{https://github.com/worldbank/d4di/tree/master/code}.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:117]:
```suggestion
To see the code examples on GitHub, go to: \url{https://github.com/worldbank/dime-data-handbook/tree/master/code}.
```

File: auxiliary/stata-guide.tex [hunk 18]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+If you are familiar with GitHub you can fork the repository and clone your fork.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:118]:
```suggestion
```

[Suggest to remove line]

File: auxiliary/stata-guide.tex [hunk 19]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+We only use Stata's built-in datasets in our code examples,
+so you do not need to download any data.
+If you have Stata installed on your computer, 
+then you will already have the data files used in the code.
+
+A less technical way to access the code is to click the individual file in the URL above, then click
+the button that says \textbf{Raw}. You will then get to a page that looks like the one at:
+\url{https://raw.githubusercontent.com/worldbank/d4di/master/code/code.do}.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:126]:
```suggestion
\url{https://raw.githubusercontent.com/worldbank/dime-data-handbook/master/code/code.do}.
```

File: auxiliary/stata-guide.tex [hunk 20]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+There, you can copy the code from your browser window to your do-file editor with the formatting intact.
+This method is only practical for a single file at the time.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:128]:
```suggestion
This method is only practical for a single file at a time.
```

File: auxiliary/stata-guide.tex [hunk 21]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+If you want to download all code used in this book, you can do that at:
+\url{https://github.com/worldbank/d4di/archive/master.zip}.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:130]:
```suggestion
\url{https://github.com/worldbank/dime-data-handbook/archive/master.zip}. 
```

File: auxiliary/stata-guide.tex [hunk 22]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+There, you can copy the code from your browser window to your do-file editor with the formatting intact.
+This method is only practical for a single file at the time.
+If you want to download all code used in this book, you can do that at:
+\url{https://github.com/worldbank/d4di/archive/master.zip}. 
+That link downloads a \texttt{.zip} file 
+with all the content used in writing this book, 
+including the \LaTeX{} code used for the book itself. 
+After extracting the .zip-file you will find all the code in a folder called \texttt{/code/}.
+
+While we try to refer to built-in commands as much as possible in this book,
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:136]:
```suggestion
While we use built-in commands as much as possible in this book,
```

File: auxiliary/stata-guide.tex [hunk 23]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+we point to user-written functions when they provide important tools.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:137]:
```suggestion
we point to user-written commands when they provide important new functionality.
```

File: auxiliary/stata-guide.tex [hunk 24]
```diff
@@ -106,32 +72,82 @@ \subsection{Writing and Understanding Stata code}
...
+In particular, we point to two suites of Stata commands developed by DIME Analytics,
+\texttt{ietoolkit}\sidenote{\url{https://dimewiki.worldbank.org/ietoolkit}} and
+\texttt{iefieldkit},\sidenote{\url{https://dimewiki.worldbank.org/iefieldkit}}
+which standardize our core data collection, management, and analysis workflows.
+In our code examples use \texttt{randtreat} or \texttt{ieboilstart}.
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:142]:
```suggestion
```

File: auxiliary/stata-guide.tex [hunk 25]
```diff
@@ -418,6 +417,12 @@ \subsection{Line breaks}
 Around that length, start a new line using \texttt{///}.
 Using \texttt{///} breaks the line in the code editor,
 while telling Stata that the same line of code continues on the next line.
+Recent version of the Stata do-file editor --
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:420]:
```suggestion
Recent versions of the Stata do-file editor --
```

File: auxiliary/stata-guide.tex [hunk 26]
```diff
@@ -418,6 +417,12 @@ \subsection{Line breaks}
...
+and many other code editors --
+automatically wraps code lines that are too long,
```
Comment (bbdaniels) [inline auxiliary/stata-guide.tex:422]:
```suggestion
automatically wrap code lines that are too long,
```

File: chapters/0-introduction.tex [hunk 27]
```diff
@@ -331,8 +331,11 @@ \section{Documenting data work with standardized code}
 due to its persistent popularity in development economics and econometrics.\sidenote{
   \url{https://aeadataeditor.github.io/presentation-20191211/\#9}}
 This book also includes, as an appendix,
-the \textbf{DIME Analytics Stata Style Guide}
-that we use in our work, which provides
+the \textbf{DIME Analytics Coding Guide} 
+which includes instructions for how to write good and not just correct code,
```
Comment (bbdaniels) [inline chapters/0-introduction.tex:335]:
```suggestion
which includes instructions for how to write good code,
```

File: chapters/0-introduction.tex [hunk 28]
```diff
@@ -331,8 +331,11 @@ \section{Documenting data work with standardized code}
...
+instructions on how to use the code examples in this book,
+as well as our Stata Style Guide.
+We use the style guide in our work, which provides
```
Comment (bbdaniels) [inline chapters/0-introduction.tex:338]:
```suggestion
We follow the style guide in our work, which provides
```

**Comment by bbdaniels** [comment]
@kbjarkefur I passed over this as well and rewrote a few bits. If you're happy with it I'm good to go! I think this came out great.

**Comment by bbdaniels** [comment]
Ok _now_ this is done correctly. Only minor changes. 👍🏼

**Comment by kbjarkefur** [comment]
@bbdaniels , seems like you after doing Stata edits in this branch then did chapter 5 and 6 here as well. you want to move them to the `ch-5` and `ch-6` branches?

**Comment by kbjarkefur** [comment]
Then I will merge this branch

**Comment by bbdaniels** [comment]
Done. I must have been doing something wrong with the branch selector in this app cause I thought I was doing it right!!

**Pull request merged**